View user details (librarians and students)

bulatovicb opened the pull request.

bulatovicb committed:
Login And Logout Librarian

bulatovicb committed:
https://matijabojanic.atlassian.net/browse/OL3-46

bulatovicb committed:
https://matijabojanic.atlassian.net/browse/OL3-37

bulatovicb committed:
https://matijabojanic.atlassian.net/browse/OL3-38

marijana-codeus commented on app/Http/Middleware/IsLibrarian.php:
```diff
@@ -0,0 +1,25 @@
+        $user = Auth::user();
+
+
+        if (!$user || $user-> role !== 'librarian') {
```
@bulatovicb Pazi na ovo, $user ->role ti nece raditi zbog space-a. Takodje, moras u User modelu definisati role relaciju da bi mogla da joj pristupis. I tu obrati paznju sta ti je id a sta ime zbog provjere vrijednosti

marijana-codeus replied:
@bulatovicb Aha tebi je role samo string nije FK. Ja bih radije da idemo sa posebnom tabelom roles i onda da ih spojimo tako sto ce users tabela ima ti role_id koji ce pokazivati na roles tabelu

bulatovicb replied:
Dodala sam novu migraciju create_roles_table
Uklonila enum i dodala role_id kao FK.
Dodala relaciju belongsTo u User model i napravila novi model Role gdje sam dodala relaciju hasMany

marijana-codeus commented on routes/api.php:
```diff
@@ -3,6 +3,9 @@
 
-Route::post('librarians', [\App\Http\Controllers\LibrarianController::class, 'store']);
-
+Route::post('create', [\App\Http\Controllers\UserController::class, 'create'])->middleware('auth:sanctum','librarian');
```
@bulatovicb Je li ideja da koristis odvojene rute za kreiranje librariana i usera?

bulatovicb replied:
Pise da samo librarian ima pravo da pravi nove usere

bulatovicb committed:
Summary promjena

- Svaki property stavljen u novi red
- Dodata nova migracija 'roles', uklonjen enum i dodat role_id kao FK
- Ispravljena greška sa 'sometimes|required'

bulatovicb committed:
Summary promjena

Dodala sam novu migraciju create_roles_table

Uklonila enum i dodala role_id kao FK.

https://matijabojanic.atlassian.net/browse/OL3-37

marijana-codeus commented on app/Http/Middleware/IsLibrarian.php:
```diff
@@ -0,0 +1,22 @@
+    {
+        $user = Auth::user();
+
+        if (!$user || $user->role_id !== 2) {
```
@bulatovicb Ovo je odlican primjer na osnovu koga mozemo da vidimo jos jedan koncept: https://refactoring.guru/replace-magic-number-with-symbolic-constant . Evo link da detaljnije procitas, ali ovo se u programiranju zovu magicni brojevi. Drugim developerima koji citaju tvoj kod (a moguce ni tebi posle nekog vremena) nije dovoljno jasno njihovo znacenje. Imas  vise opcija, mozes da dodas dvije konstante u Role modelu i da zamijenis 2 recimo sa Role::LIBRARIAN, da napravis funkciju isLibrarian, i sl.

bulatovicb committed:
Added magic numbers in User model

Added comments on functions

https://matijabojanic.atlassian.net/browse/OL3-37

marijana-codeus commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,265 @@
+        if ($request->filled('search_value')) {
+            $search = strtolower($request->search_value);
+            $query->where(function ($q) use ($search) {
+                $q->whereRaw('LOWER(first_name) LIKE ?', ["%$search%"])
```
@bulatovicb Treba nam ILIKE umjesto LIKE

marijana-codeus commented on routes/api.php:
```diff
@@ -1,8 +1,16 @@
+Route::post('logout', [\App\Http\Controllers\AuthController::class, 'logout'])->middleware('auth:sanctum');
 
-Route::post('librarians', [\App\Http\Controllers\LibrarianController::class, 'store']);
+Route::post('create', [\App\Http\Controllers\UserController::class, 'create'])->middleware('auth:sanctum', 'librarian');
```
@bulatovicb Posto ti svi ovi endpointi zahtijevaju autorizaciju, da ne bi morala da ponavljas svaki put, mozes i da ih grupises:
```
Route::group(['middleware' => 'auth:sanctum'], function () {

}
```

marijana-codeus commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,265 @@
+            'username' => $user->username,
+            'jmbg' => $user->jmbg,
+            'role' => $user->role->name,
+            'profile_picture_url' => $user->profile_picture
```
@bulatovicb mozes ukloniti ovo slobodno, dovoljno je da imas posebnu rutu, FE ce znati sta da poziva

marijana-codeus commented on routes/api.php:
```diff
@@ -1,8 +1,16 @@
 
-Route::post('librarians', [\App\Http\Controllers\LibrarianController::class, 'store']);
+Route::post('create', [\App\Http\Controllers\UserController::class, 'create'])->middleware('auth:sanctum', 'librarian');
+Route::get('users/{username}', [\App\Http\Controllers\UserController::class, 'show'])->middleware('auth:sanctum', 'librarian');
```
@bulatovicb Baci pogled na ovo: https://laravel.com/docs/12.x/routing#route-model-binding. Ovo je jedna olaksica koju Laravel nudi, mozda ne bi bilo lose da je istrazis. Po defaultu se binding radi preko id kolone, medjutim ovdje ti je objasnjeno i kako ti mozes da zadas custom key (sto bi u tvom slucaju bio username). Mada ja svakako ne bih bjezala od koriscenja id-ja ovdje, osim ako nemas neki specifican razlog zasto ti treba bas username

marijana-codeus commented on app/Http/Controllers/UserController.php:
```diff
@@ -0,0 +1,265 @@
+        $user = Auth::user();
+
+        $validator = Validator::make($request->all(), [
+            'profile_picture' => 'nullable|image|max:5120',
```
@bulatovicb Posto je ovo ruta koja sluzi samo za update slike, imalo bi mi smisla je profile_picture required

marijana-codeus commented on app/Models/User.php:
```diff
@@ -5,41 +5,49 @@
+        'first_name',
+        'last_name',
+        'username',
+        'role',
```
@bulatovicb Role bi trebalo obrisati posto smo dodali role_id

bulatovicb committed:
feat(OL3-37): Poboljšanja korisničkog upravljanja i API ruta

Zamijenjen LIKE sa ILIKE za case-insensitive pretragu korisnika

Primijenjeno grupisanje ruta

Omogućen Route Model Binding koristeći 'username' umjesto ID

U metodi updateProfilePicture, 'profile_picture' polje je sada required

Uklonjen zastarjeli 'role' property iz User modela

https://matijabojanic.atlassian.net/browse/OL3-37

bulatovicb committed:
Izmjena funkcije destroy

https://matijabojanic.atlassian.net/browse/OL3-37
